### DISCUSSION TITLE
Is it possible to auto generate docusarus sidebars from i18n folder?

### DISCUSSION THREAD

**Discussion by isorna** [facebook/docusaurus#7585] [Q&A]
As far as I understand, sidebars can only be auto generated from the /docs folder, and as I'm using docusaurus v2 the docs translations will allow me to use different filenames and slugs, but it's impossible to assign sidebars dirName to a folder outside the `/docs` folder:

```javascript
const sidebars = { 
    i18nSidebars: [
        {
            type: 'autogenerated',
            dirName: 'i18n/[locale]/docusaurus-plugin-content-docs/[version]'
        }
    ]
};
module.exports = sidebars;
```

There's no information about this on the docusaurus documentation (no interaction between content-docs plugin's sidebars and i18n), I've tried different configurations with no success. When I try to set `dirName` outside the `/docs` folder I always get the same warning:

`[WARNING] No docs found in "../i18n/es/docusaurus-plugin-content-docs/current": can't auto-generate a sidebar.`

The problem I see is that once I started using Docusaurus i18n support, I've to copy `/docs` folder content inside `/i18n/[lang]/docusaurus-plugin-content-docs/[version]` to get the translated content built, but the autogenerated sidebars are built from the `/docs` folder content, so I'm forced to mantain several copies of the same content structure but with the same slugs and filenames (those of the `/docs` folder), ignoring the translated content structure and `_category_.json`.

I could create different sidebars and use the `displayed_sidebar` parameter to switch them from the i18n translated docs, but that also means an extra step on the mantainment of my project as I've one git submodule for each documentation section inside the project, so that each section can be translated by each translation team.

This way the app main repository is not affected by any submodule repository, and controls the deploy process.

**Comment by slorber** [comment]
Docusaurus somehow expects all the localized versions of your site to have the same structure:
- same doc set (localized or not)
- same urls/slug

This is important to ensure our i18n system works fine. 
Otherwise one site would have to know the localized URLs of all the other sites for each docs (to switch language through the dropdown, or to generate SEO hreflang headers)

---

`displayed_sidebar` forces the display of a given sidebar while browsing a doc, but it does not prevent a doc to appear if it's found in the autogenerated folder. IE `displayed_sidebar: null` means browsing the doc shows no sidebar, but browsing other docs pages may still display the doc in the sidebar. (not easy to explain 🤪 )

That being said you could be able to customize the sidebar per locale by using a custom generator: https://docusaurus.io/docs/sidebar/autogenerated#customize-the-sidebar-items-generator

--- 

It's hard to help you figure out the best solution because we are discussing technical details and I still don't understand the business outcome you try to achieve. Please start with why :)

**Reply by isorna** [reply]
Hi Sébastien, thanks for the update.
I'll try to make clear my needs.

I'm workong on a pwa that will include the documentation from several projects (as well as some client side utilities for the user).

Each documented project will be translated and updated  by different localized teams, and although the original documentation is written in english, I've no control on each of the teams (only the spanish one), as each country organizes by itself.

So I thought that including all the documentation projects inside the main pwa was going to be impossible, then I moved each doc project into a separate git submodule, where the main branch is the en-US language and there's a branch for each other language, for each documentation project.

This allows me to "control" the i18n folder, as I can link each project with the needed branch inside the i18n docs plugin current folder. Once the CI launches, it checks out each submodule/branch and builds the whole pwa before publishing it.

But then I found that I've to also check out the main branch of each project inside the /docs folder, just only for the document's sidebar  structure and that seemed quite too much for such a simple task, as you said docusaurus expects to maintain the same structure inside the i18n docs plugin folder, so I (wrongly) assumed that I could assign an autogenerated folder structure pointing to the i18n/[language]/ docs plugin folder, and that's what I was asking for.

I know that it's not a simple task, but it may really help to translate also the slugs and the sidebar structure, as not all languages share the same structure, and it's really weird to read a documentation in one language and see other in the url.

Regards

**Reply by slorber** [reply]
It's very hard to understand 🤪 a repro/example would be more helpful

> as you said docusaurus expects to maintain the same structure inside the i18n docs plugin folder, so I (wrongly) assumed that I could assign an autogenerated folder structure pointing to the i18n/[language]/ docs plugin folder, and that's what I was asking for.

We always use the source docs (untranslated) to create the autogenerated sidebar, not the localized folder.

This is done on purpose to ensure that all languages always have the exact same number of pages.

> I know that it's not a simple task, but it may really help to translate also the slugs and the sidebar structure, as not all languages share the same structure, and it's really weird to read a documentation in one language and see other in the URL.

This is very unlikely we change this in the near future. 

You can already translate slug frontmatter in your docs, but this would break:
- SEO hreflang metadata
- the ability to switch from one language to another in the navbar, while staying on the same document (the dropdown would have to know all the URLs, of all the locales for that to work => heavy)

For the sidebar structure, I don't really understand why a localized site would have a different one. Something you could try is to add "draft: true" in localized docs if you don't want some of them to appear, but it may not be what you are looking for. (and it would still break the 2 things above)

---

We have built an opinionated i18n and versioning system, now it's up to you to use it or not.

You can as well deploy one site per version/locale, like most people used to do before, using Git branches...

You'll find interesting i18n strategies on the initial i18n RFC here: https://github.com/facebook/docusaurus/issues/3317

For example, the React website has an interesting strategy where each localized site is a Git fork of the upstream website. (the React/S old site, will be replaced soon by a new one)

**Reply by isorna** [reply]
Ok thanks, I'll think about a solution for my particular use case.
Regards

**Comment by isorna** [comment]
Hi Sébastien, thanks for the update.
I'll try to make clear my needs.

I'm workong on a pwa that will include the documentation from several
projects (as well as some client side utilities for the user).

Each documented project will be translated and updated  by different
localized teams, and although the original documentation is written in
english, I've no control on each of the teams (only the spanish one), as
each country organizes by itself.

So I thought that including all the documentation projects inside the main
pwa was going to be impossible, then I moved each doc project into a
separate git submodule, where the main branch is the en-US language and
there's a branch for each other language, for each documentation project.

This allows me to "control" the i18n folder, as I can link each project
with the needed branch inside the i18n docs plugin current folder. Once the
CI launches, it checks out each submodule/branch and builds the whole pwa
before publishing it.

But then I found that I've to also check out the main branch of each
project inside the /docs folder, just only for the document's sidebar
structure and that seemed quite too much for such a simple task, as you
said docusaurus expects to maintain the same structure inside the i18n docs
plugin folder, so I (wrongly) assumed that I could assign an autogenerated
folder structure pointing to the i18n/[language]/ docs plugin folder, and
that's what I was asking for.

I know that it's not a simple task, but it may really help to translate
also the slugs and the sidebar structure, as not all languages share the
same structure, and it's really weird to read a documentation in one
language and see other in the url.

Regards

**Reply by slorber** [reply]
> I know that it's not a simple task, but it may really help to translate
also the slugs and the sidebar structure, as not all languages share the
same structure, and it's really weird to read a documentation in one
language and see other in the URL.

I understand there's value in this but it's a complex feature that only very limited users are needing, and it's really not a top priority for us to implement it.

To explain more the complexity, we want:
- the locale dropdown to be able to switch from one locale to another without losing the "context" (ie not display 404 when switching because a slug is localized)
- show correct hreflang headers in all pages for SEO reasons

This meta is present on our pages:

<img width="580" alt="CleanShot 2022-10-12 at 11 23 40@2x" src="https://user-images.githubusercontent.com/749374/195304785-3ec05e0e-49ba-4ae8-9caa-e49980fa652d.png">

To make this work with localized slugs, and localized sites potentially containing more/less pages than the original site, you'd need the site to be aware of all the URLs of all the locales you plan to build into.

This means that when building the French site, you'd have to read markdown files for ALL other languages to figure out the final URL of the same page in all the other locales. 

That's another scalability problem to put on the table if you need to spend `markdown reading time * number of locales`. And if some localized sites may contain less pages, it's even more difficult to be able to read.

**Reply by slorber** [reply]
Regarding your question: 

> Is it possible to auto generate docusarus sidebars from i18n folder?

Technically you can write your own custom sidebar item generator and read a folder in `i18n` instead of `docs`. Maybe you can also use a locale env variable in the `sidebars.js` file to create a different tree based on the locale

We don't prevent you from localizing slugs atm, but we don't recommend doing this because we don't provide any logic for seamless dropdown language switch or hreflang support.

But:
- you can swizzle the theme to bring your own hreflang SEO support
- you can implement your own locale dropdown supporting switching to localized slugs (https://github.com/facebook/docusaurus/issues/7227)

In other words: you can probably implement what you want in userland today, but yes it's definitively not easy.
